File: translator.py
import os
import requests
import logging
import subprocess
from constants import (
    OLLAMA_URL, OLLAMA_TAGS_URL, MODEL, TRANSLATE_TEXT_PROMPT, API_KEY,
    AGY_PATH, PREFERRED_PROVIDER
)

# Set up logging to both terminal and file
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
    handlers=[
        logging.FileHandler("log.log"),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger("Translator")

def translate_with_antigravity(prompt: str) -> str:
    """Executes prompt non-interactively using Antigravity (agy -p)."""
    try:
        logger.info(f"Executing prompt via Antigravity CLI ({AGY_PATH} -p)...")
        result = subprocess.run(
            [AGY_PATH, "-p", prompt],
            capture_output=True, text=True, timeout=120
        )
        if result.returncode == 0 and result.stdout.strip():
            logger.info("Antigravity request succeeded.")
            return result.stdout.strip()
        else:
            err_msg = result.stderr.strip() or "Empty stdout returned"
            logger.error(f"Antigravity CLI error (code {result.returncode}): {err_msg}")
            raise RuntimeError(f"Antigravity error: {err_msg}")
    except Exception as e:
        logger.error(f"Antigravity request failed ({type(e).__name__}: {e})")
        raise

def get_available_ollama_model() -> str:
    """Gets configured model or picks first available installed model from Ollama."""
    try:
        resp = requests.get(OLLAMA_TAGS_URL, timeout=5)
        if resp.status_code == 200:
            models = [m.get("name") for m in resp.json().get("models", []) if m.get("name")]
            if MODEL in models or f"{MODEL}:latest" in models:
                return MODEL
            if models:
                gen_models = [m for m in models if "embed" not in m.lower()]
                selected = gen_models[0] if gen_models else models[0]
                logger.info(f"Model '{MODEL}' not found on Ollama. Using available model '{selected}'")
                return selected
    except Exception as e:
        logger.warning(f"Could not query Ollama tags endpoint: {e}")
    return MODEL

def translate_with_ollama(prompt: str) -> str:
    """Translates prompt using local or cloud Ollama API instance."""
    model_name = get_available_ollama_model()
    logger.info(f"Starting translation request to {OLLAMA_URL} with model '{model_name}'")
    
    headers = {"Content-Type": "application/json"}
    if API_KEY:
        headers["Authorization"] = f"Bearer {API_KEY}"
    
    payload = {"model": model_name, "prompt": prompt, "stream": False}
    
    response = requests.post(
        OLLAMA_URL,
        json=payload,
        headers=headers,
        timeout=120
    )
    logger.info(f"Ollama responded with HTTP {response.status_code}")
    response.raise_for_status()
    
    result = response.json().get("response", "").strip()
    if not result:
        raise ValueError("Ollama returned an empty response.")
    logger.info("Successfully extracted Ollama response.")
    return result

def translate_text(text: str, is_raw_prompt: bool = False) -> str:
    """
    Translates or solves text using configured provider with automatic fallback.
    If PREFERRED_PROVIDER is 'antigravity', tries agy -p first then Ollama.
    If 'ollama', tries Ollama first then agy -p.
    """
    prompt = text if is_raw_prompt else TRANSLATE_TEXT_PROMPT.format(text=text)
    
    if PREFERRED_PROVIDER == "ollama":
        providers = [("Ollama", translate_with_ollama), ("Antigravity", translate_with_antigravity)]
    else:  # 'antigravity' or 'auto'
        providers = [("Antigravity", translate_with_antigravity), ("Ollama", translate_with_ollama)]
        
    last_error = None
    for provider_name, provider_fn in providers:
        try:
            logger.info(f"Attempting request using {provider_name}...")
            return provider_fn(prompt)
        except Exception as e:
            logger.warning(f"{provider_name} failed: {e}. Trying fallback if available...")
            last_error = e
            
    return f"Execution failed on all available providers. Error: {last_error}"

def ask_antigravity(question: str) -> str:
    """Direct helper for querying Antigravity CLI."""
    return translate_text(question, is_raw_prompt=True)

def translate_image(image_data):
    """Placeholder for image translation."""
    return "Image translation placeholder. (Vision API not yet connected)"

File: translator.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

var (
	preferredProvider = getEnvOrDefault("OVERLAY_PROVIDER", "ollama")
	ollamaURL         = getEnvOrDefault("OLLAMA_URL", "http://localhost:11434/api/generate")
	ollamaTagsURL     = getEnvOrDefault("OLLAMA_TAGS_URL", "http://localhost:11434/api/tags")
	ollamaModel       = getEnvOrDefault("OLLAMA_MODEL", "gemini-3.1-pro-high")
	ollamaAPIKey      = os.Getenv("OLLAMA_API_KEY")

	agyModel  = getEnvOrDefault("AGY_MODEL", "gemini-3.1-pro-high")
	agyEffort = getEnvOrDefault("AGY_EFFORT", "high")
	agyPath   = resolveAgyPath()
)

const translateTextPrompt = `Here is my coding challenge problem. Act like a candidate in a technical interview.

Provide:
1. A simple, clean, and easy-to-understand solution (nothing fancy, use basic code).
2. A short, clear, and user-friendly explanation as you would explain to an interviewer.

Problem:
%s`

func getEnvOrDefault(key, fallback string) string {
	if val := os.Getenv(key); val != "" {
		return val
	}
	return fallback
}

func resolveAgyPath() string {
	if p, err := exec.LookPath("agy"); err == nil {
		return p
	}
	homeAgy := filepath.Join(os.Getenv("HOME"), ".local", "bin", "agy")
	if _, err := os.Stat(homeAgy); err == nil {
		return homeAgy
	}
	return "agy"
}

type OllamaTagsResponse struct {
	Models []struct {
		Name string `json: "name"`
	} `json:"models"`
}

type OllamaGenerateRequest struct {
	Model  string `json:"model"`
	Prompt string `json:"prompt"`
	Stream bool   `json:"stream"`
}

type OllamaGenerateResponse struct {
	Response string `json:"response"`
}

func getAvailableOllamaModel() string {
	client := http.Client{Timeout: 5 * time.Second}
	resp, err := client.Get(ollamaTagsURL)
	if err == nil && resp.StatusCode == 200 {
		defer resp.Body.Close()
		var tags OllamaTagsResponse
		if err := json.NewDecoder(resp.Body).Decode(&tags); err == nil {
			var names []string
			for _, m := range tags.Models {
				if m.Name != "" {
					names = append(names, m.Name)
					if m.Name == ollamaModel || m.Name == ollamaModel+":latest" {
						return ollamaModel
					}
				}
			}
			for _, name := range names {
				if !strings.Contains(strings.ToLower(name), "embed") {
					log.Printf("Model '%s' not found on Ollama. Using available model '%s'", ollamaModel, name)
					return name
				}
			}
			if len(names) > 0 {
				return names[0]
			}
		}
	}
	return ollamaModel
}

func translateWithOllama(prompt string) (string, error) {
	modelName := getAvailableOllamaModel()
	log.Printf("Starting translation request to %s with model '%s'", ollamaURL, modelName)

	reqBody, err := json.Marshal(OllamaGenerateRequest{
		Model:  modelName,
		Prompt: prompt,
		Stream: false,
	})
	if err != nil {
		return "", err
	}

	req, err := http.NewRequest("POST", ollamaURL, bytes.NewBuffer(reqBody))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	if ollamaAPIKey != "" {
		req.Header.Set("Authorization", "Bearer "+ollamaAPIKey)
	}

	client := http.Client{Timeout: 120 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("Ollama returned HTTP %d: %s", resp.StatusCode, string(body))
	}

	var genResp OllamaGenerateResponse
	if err := json.NewDecoder(resp.Body).Decode(&genResp); err != nil {
		return "", err
	}

	result := strings.TrimSpace(genResp.Response)
	if result == "" {
		return "", fmt.Errorf("Ollama returned an empty response")
	}
	return result, nil
}

func translateWithAntigravity(prompt string) (string, error) {
	args := []string{}
	if agyModel != "" {
		args = append(args, "--model", agyModel)
	}
	if agyEffort != "" {
		args = append(args, "--effort", agyEffort)
	}
	args = append(args, "-p", prompt)

	cmd := exec.Command(agyPath, args...)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	log.Printf("Executing prompt via Antigravity CLI (%s)...", agyPath)
	err := cmd.Run()
	if err == nil && strings.TrimSpace(stdout.String()) != "" {
		return strings.TrimSpace(stdout.String()), nil
	}

	errMsg := strings.TrimSpace(stderr.String())
	if errMsg == "" && err != nil {
		errMsg = err.Error()
	}
	return "", fmt.Errorf("Antigravity error: %s", errMsg)
}

func translateText(text string, isRawPrompt bool) string {
	prompt := text
	if !isRawPrompt {
		prompt = fmt.Sprintf(translateTextPrompt, text)
	}

	type provider struct {
		name string
		fn   func(string) (string, error)
	}

	var providers []provider
	if preferredProvider == "ollama" {
		providers = []provider{
			{"Ollama", translateWithOllama},
			{"Antigravity", translateWithAntigravity},
		}
	} else {
		providers = []provider{
			{"Antigravity", translateWithAntigravity},
			{"Ollama", translateWithOllama},
		}
	}

	var lastErr error
	for _, p := range providers {
		log.Printf("Attempting request using %s...", p.name)
		result, err := p.fn(prompt)
		if err == nil {
			return result
		}
		log.Printf("%s failed: %v. Trying fallback...", p.name, err)
		lastErr = err
	}

	return fmt.Sprintf("Execution failed on all available providers. Error: %v", lastErr)
}

func askAntigravity(question string) string {
	result, err := translateWithAntigravity(question)
	if err != nil {
		return fmt.Sprintf("Antigravity error: %v", err)
	}
	return result
}
